spec/services/reference_number_creator_spec.rb

```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
+RSpec.describe ReferenceNumberCreator do
+  subject { ReferenceNumberCreator.new }
+
+  describe '#call' do
+    it 'matches the regex' do
+      expect(subject.call).to match(/L(-[ABCDEFHJKLMNPRTUVWXY0-9]{3}){2}/)
+    end
+
+    context 'first generated numbers are taken' do
+      let(:existing_reference_number) { 'L-AAA-AAA' }
+      let(:other_existing_reference_number) { 'L-BBB-BBB' }
+      let(:third_reference_number) { 'L-CCC-CCC' }
+
+      before do
+        (create :legal_aid_application).update!(application_ref: existing_reference_number)
+        (create :legal_aid_application).update!(application_ref: other_existing_reference_number)
```

Comment:
Creating and then updating is unnecessary I think as the `create_app_ref` is irrelevant as the ref is being supplied. As the update is not required, the creation can be done in a `let!` rather than using a before. So:
```ruby
let!(:legal_aid_application) { create :legal_aid_application, application_ref: existing_reference_number }
let!(:other_legal_aid_application) { create :legal_aid_application, application_ref: other_existing_reference_number }
```

Reply:
I tried your solution but the problem is that `application_ref` get overwritten with a random reference number because of the `before_create :create_app_ref` callback

Reply:
Then doesn't `legal_aid_application#create_app_ref` need updating so that it doesn't behave that way (as the current implementation has the potential to create a lot of unexpected behaviour):
```ruby
  def create_app_ref
    self.application_ref = SecureRandom.uuid unless application_ref.present?
  end
```

Reply:
Yes good idea

Reply:
And of course it should now be:
```ruby
self.application_ref = ReferenceNumberCreator.call unless application_ref.present?
```